lib/Core/PTree.cpp

```diff
@@ -25,9 +46,23 @@ PTree::split(Node *n,
   assert(n && !n->left && !n->right);
   n->left = new Node(n, leftData);
   n->right = new Node(n, rightData);
+
+  if (dumpTree)
+    printChildrenToCsv(n);
+
   return std::make_pair(n->left, n->right);
 }
 
+void PTree::printChildrenToCsv(Node *parent) {
```

Comment:
Why is this so tied to printing? I think it would be better to have a general map/iterator over the tree and then do the printing in the stats tracker. I don't think introducing this new stats dependency is needed or useful.

Reply:
> Why is this so tied to printing? I think it would be better to have a general map/iterator over the tree and then do the printing in the stats tracker. I don't think introducing this new stats dependency is needed or useful.

Wouldn't this involve having to iterate over the entire tree every time we want to convert it into a CSV? With the current method, we only need to handle the current node

Reply:
At the moment we print out each new link between parent and child node when the nodes are created. This means that if the tree is pruned at any point then we will still have the whole tree printed.

For our use case it's very unlikely that we'll ever be interested in a tree that is big enough to need pruning but it still seems like it would be nice to print the whole tree regardless.

How would it be if we added something like addSplitListener() to PTree, this would take a callback (that accepts the new parent node and its children) as a parameter?

Let me know if that isn't clear :)

Reply:
> be interested in a tree that is big enough to need pruning

That happens automatically as soon as a branch terminates.

> we added something like addSplitListener

I wouldn't over-engineer it. Currently that's the only use case.

In general I would also remove the dependency on StatsTracker. Pass a file reference to the constructor or to a new register method.

Reply:
Ok I see. You could just block iteration and the have a seperate thread continously writting out this CSV, but it's perhaps better not to go there.

The listener/callback approach sounds promising. But if you are doing that you might as well do a forking callback, which would have 2 uses: 1 for building the PTree and one for writting out the CSV. 

Maybe that's a bit complicated, I just really don't like the StatsTracker in PTree.